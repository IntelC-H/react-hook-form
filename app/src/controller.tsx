import React from 'react';
import { useForm, Controller } from 'react-hook-form';
import ReactSelect from 'react-select';
import {
  TextField,
  Checkbox,
  Select,
  MenuItem,
  Switch,
  RadioGroup,
  FormControlLabel,
  Radio,
} from '@material-ui/core';

let renderCount = 0;

const options = [
  { value: 'chocolate', label: 'Chocolate' },
  { value: 'strawberry', label: 'Strawberry' },
  { value: 'vanilla', label: 'Vanilla' },
] as any;

const defaultValues = {
  Native: '',
  TextField: '',
  Select: '',
  ReactSelect: '',
  Checkbox: false,
  switch: false,
  RadioGroup: '',
};

type Form = {
  Native: string;
  TextField: string;
  Select: string;
  ReactSelect: string;
  Checkbox: boolean;
  switch: boolean;
  RadioGroup: string;
};

export default function Field(props: any) {
  const methods = useForm<Form>({
    defaultValues,
    mode: props.match.params.mode,
  });
  const { handleSubmit, errors, reset, control } = methods;

  renderCount++;

  return (
    <form onSubmit={handleSubmit(() => {})}>
      <div className="container">
        <section id="input-checkbox">
          <label>MUI Checkbox</label>
          <Controller
            as={<Checkbox />}
            name="Checkbox"
            control={control}
            rules={{ required: true }}
          />
        </section>

        {errors.Checkbox && <p id={'Checkbox'}>Checkbox Error</p>}

        <section id="input-radio-group">
          <label>Radio Group</label>
          <Controller
            render={(props) => (
              <RadioGroup aria-label="gender" name="gender1" {...props}>
                <FormControlLabel
                  value="female"
                  control={<Radio />}
                  label="Female"
                />
                <FormControlLabel
                  value="male"
                  control={<Radio />}
                  label="Male"
                />
              </RadioGroup>
            )}
            rules={{ required: true }}
            name="RadioGroup"
            control={control}
          />
        </section>

        {errors.RadioGroup && <p id={'RadioGroup'}>RadioGroup Error</p>}

        <section id="input-textField">
          <label>MUI TextField</label>
          <Controller
            as={<TextField />}
            name="TextField"
            control={control}
            rules={{ required: true }}
          />
        </section>

        {errors.TextField && <p id="TextField">TextField Error</p>}

        <section id="input-select">
          <label>MUI Select</label>
          <Controller
            render={(props) => (
              <Select {...props}>
                <MenuItem value={10}>Ten</MenuItem>
                <MenuItem value={20}>Twenty</MenuItem>
                <MenuItem value={30}>Thirty</MenuItem>
              </Select>
            )}
            rules={{ required: true }}
            name="Select"
            control={control}
          />
        </section>

        {errors.Select && <p id="Select">Select Error</p>}

        <section id="input-switch">
          <label>MUI Switch</label>
          <Controller
            as={<Switch value="checkedA" />}
            name="switch"
            rules={{ required: true }}
            control={control}
          />
        </section>

        {errors.switch && <p id="switch">switch Error</p>}

        <section id="input-ReactSelect">
          <label>React Select</label>
          <Controller
            render={(props) => (
              <ReactSelect isClearable options={options} {...props} />
            )}
            name="ReactSelect"
            control={control}
            rules={{ required: true }}
          />
        </section>

        {errors.ReactSelect && <p id="ReactSelect">ReactSelect Error</p>}
      </div>

      <span id="renderCount">{renderCount}</span>

      <button
        type="button"
        onClick={() => {
          reset({
            Native: '',
            TextField: '',
            Select: '',
            ReactSelect: '',
            Checkbox: false,
            switch: false,
            RadioGroup: '',
          });
        }}
      >
        Reset Form
      </button>
      <button id="submit">submit</button>
    </form>
  );
}
