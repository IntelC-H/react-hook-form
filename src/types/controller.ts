import * as React from 'react';

import { RegisterOptions } from './validator';
import {
  Control,
  ControllerFieldStateError,
  FieldPath,
  FieldPathValue,
  FieldValues,
  Noop,
  RefCallBack,
  UnpackNestedValue,
  UseFormStateReturn,
} from './';

export type ControllerFieldState<
  TFieldValues extends FieldValues = FieldValues,
  TFieldName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
> = {
  invalid: boolean;
  isTouched: boolean;
  isDirty: boolean;
  error?: ControllerFieldStateError<FieldPathValue<TFieldValues, TFieldName>>;
};

export type ControllerRenderProps<
  TFieldValues extends FieldValues = FieldValues,
  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
> = {
  onChange: (...event: any[]) => void;
  onBlur: Noop;
  value: UnpackNestedValue<FieldPathValue<TFieldValues, TName>>;
  name: TName;
  ref: RefCallBack;
};

export type UseControllerProps<
  TFieldValues extends FieldValues = FieldValues,
  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
> = {
  name: TName;
  rules?: Omit<
    RegisterOptions<TFieldValues, TName>,
    'valueAsNumber' | 'valueAsDate' | 'setValueAs' | 'disabled'
  >;
  shouldUnregister?: boolean;
  defaultValue?: UnpackNestedValue<FieldPathValue<TFieldValues, TName>>;
  control?: Control<TFieldValues>;
};

export type UseControllerReturn<
  TFieldValues extends FieldValues = FieldValues,
  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
> = {
  field: ControllerRenderProps<TFieldValues, TName>;
  formState: UseFormStateReturn<TFieldValues>;
  fieldState: ControllerFieldState<TFieldValues, TName>;
};

export type ControllerProps<
  TFieldValues extends FieldValues = FieldValues,
  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
> = {
  render: ({
    field,
    fieldState,
    formState,
  }: {
    field: ControllerRenderProps<TFieldValues, TName>;
    fieldState: ControllerFieldState<TFieldValues, TName>;
    formState: UseFormStateReturn<TFieldValues>;
  }) => React.ReactElement;
} & UseControllerProps<TFieldValues, TName>;
