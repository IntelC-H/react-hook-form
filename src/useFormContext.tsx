import * as React from 'react';
import { FormContextValues, FormProps } from './types';

export const FormGlobalContext = React.createContext<FormContextValues>(
  {} as FormContextValues,
);

export const useFormContext = () =>
  React.useContext<FormContextValues>(FormGlobalContext);

export function FormContext(props: FormProps) {
  const { children, ...rest } = props;
  return (
    <FormGlobalContext.Provider
      value={
        {
          ...rest,
        } as FormContextValues
      }
    >{children}</FormGlobalContext.Provider>
    
  );
}
