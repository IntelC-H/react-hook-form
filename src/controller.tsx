import * as React from 'react';
import isBoolean from './utils/isBoolean';
import isUndefined from './utils/isUndefined';
import get from './utils/get';
import set from './utils/set';
import getInputValue from './logic/getInputValue';
import skipValidation from './logic/skipValidation';
import isNameInFieldArray from './logic/isNameInFieldArray';
import { useFormContext } from './useFormContext';
import { VALUE } from './constants';
import { Control, Field } from './types/form';
import { ControllerProps } from './types/props';

const Controller = <
  TAs extends
    | React.ReactElement
    | React.ComponentType<any>
    | keyof JSX.IntrinsicElements,
  TControl extends Control = Control
>({
  name,
  rules,
  as,
  render,
  defaultValue,
  control,
  onFocus,
  ...rest
}: ControllerProps<TAs, TControl>) => {
  const methods = useFormContext();
  const {
    defaultValuesRef,
    setValue,
    register,
    unregister,
    errorsRef,
    removeFieldEventListener,
    trigger,
    mode: { isOnSubmit, isOnBlur, isOnChange },
    reValidateMode: { isReValidateOnBlur, isReValidateOnSubmit },
    isSubmittedRef,
    touchedFieldsRef,
    readFormStateRef,
    reRender,
    fieldsRef,
    fieldArrayNamesRef,
  } = control || methods.control;
  const [value, setInputStateValue] = React.useState(
    isUndefined(defaultValue)
      ? get(defaultValuesRef.current, name)
      : defaultValue,
  );
  const valueRef = React.useRef(value);
  const isCheckboxInput = isBoolean(value);
  const shouldReValidateOnBlur = isOnBlur || isReValidateOnBlur;
  const onFocusRef = React.useRef(onFocus);
  const isNotFieldArray = !isNameInFieldArray(fieldArrayNamesRef.current, name);
  const isSubmitted = isSubmittedRef.current;

  const shouldValidate = () =>
    !skipValidation({
      hasError: !!get(errorsRef.current, name),
      isOnBlur,
      isOnSubmit,
      isOnChange,
      isReValidateOnBlur,
      isReValidateOnSubmit,
      isSubmitted,
    });

  const commonTask = (event: any[]) => {
    const data = getInputValue(event[0], isCheckboxInput);
    setInputStateValue(data);
    valueRef.current = data;
    return data;
  };

  const registerField = React.useCallback(() => {
    if (fieldsRef.current[name]) {
      fieldsRef.current[name] = {
        ref: fieldsRef.current[name]!.ref,
        ...rules,
      };
    } else {
      if (!isNotFieldArray) {
        removeFieldEventListener(fieldsRef.current[name] as Field, true);
      }

      register(
        Object.defineProperty({ name, focus: onFocusRef.current }, VALUE, {
          set(data) {
            setInputStateValue(data);
            valueRef.current = data;
          },
          get() {
            return valueRef.current;
          },
        }),
        rules,
      );
    }
  }, [
    isNotFieldArray,
    fieldsRef,
    rules,
    name,
    onFocusRef,
    register,
    removeFieldEventListener,
  ]);

  React.useEffect(
    () => () => {
      !isNameInFieldArray(fieldArrayNamesRef.current, name) && unregister(name);
    },
    [unregister, name, fieldArrayNamesRef],
  );

  React.useLayoutEffect(() => {
    registerField();
  }, [registerField]);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  React.useLayoutEffect(() => {
    if (!fieldsRef.current[name]) {
      registerField();
      if (isNotFieldArray) {
        setInputStateValue(
          isUndefined(defaultValue)
            ? get(defaultValuesRef.current, name)
            : defaultValue,
        );
      }
    }
  });

  const onBlur = () => {
    if (
      readFormStateRef.current.touched &&
      !get(touchedFieldsRef.current, name)
    ) {
      set(touchedFieldsRef.current, name, true);
      reRender();
    }

    if (shouldReValidateOnBlur) {
      trigger(name);
    }
  };

  const onChange = (...event: any[]) =>
    setValue(name, commonTask(event), shouldValidate());

  const handlerProps = {
    onChange: React.useCallback(onChange, []),
    onBlur: React.useCallback(onBlur, []),
  };

  const props = {
    ...rest,
    ...handlerProps,
    ...{ [isCheckboxInput ? 'checked' : VALUE]: value },
  };

  return as
    ? React.isValidElement(as)
      ? React.cloneElement(as, props)
      : React.createElement(as as string, props)
    : render
    ? render({
        ...handlerProps,
        value,
      })
    : null;
};

export { Controller };
