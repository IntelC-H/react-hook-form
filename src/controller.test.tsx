import * as React from 'react';
import { render, fireEvent, act, screen } from '@testing-library/react';
import { Controller } from './controller';
import { FormProvider } from './useFormContext';
import { useFieldArray } from './useFieldArray';
import { useForm } from './useForm';

const Input = ({ onChange, onBlur, placeholder }: any) => (
  <input
    placeholder={placeholder}
    onChange={() => onChange?.(1, 2)}
    onBlur={() => onBlur?.(1, 2)}
  />
);

describe('Controller', () => {
  it('should render correctly with as with string', () => {
    const Component = () => {
      const { control } = useForm();
      return (
        <Controller
          defaultValue=""
          name="test"
          as={'input' as const}
          control={control}
        />
      );
    };

    render(<Component />);

    const input = screen.queryByRole('textbox') as HTMLInputElement | null;

    expect(input).toBeInTheDocument();
    expect(input?.name).toBe('test');
  });

  it('should render correctly with as with component', () => {
    const Component = () => {
      const { control } = useForm();
      return (
        <Controller
          defaultValue=""
          name="test"
          as={<input />}
          control={control}
        />
      );
    };

    render(<Component />);

    const input = screen.queryByRole('textbox') as HTMLInputElement | null;

    expect(input).toBeInTheDocument();
    expect(input?.name).toBe('test');
  });

  it('should be assigned method.control variable if wrap with FormProvider', () => {
    const Component = () => {
      const methods = useForm();
      return (
        <FormProvider {...methods}>
          <Controller defaultValue="" name="test" as={'input' as const} />
        </FormProvider>
      );
    };

    expect(() => render(<Component />)).not.toThrow();
  });

  it('should reset value', async () => {
    const Component = () => {
      const { reset, control } = useForm();

      return (
        <>
          <Controller
            defaultValue="default"
            name="test"
            as={<input />}
            control={control}
          />
          <button type="button" onClick={() => reset()}>
            reset
          </button>
        </>
      );
    };

    render(<Component />);

    fireEvent.input(screen.getByRole('textbox'), { target: { value: 'test' } });
    expect(screen.getByRole('textbox')).toHaveValue('test');

    fireEvent.click(screen.getByRole('button', { name: /reset/i }));
    expect(screen.getByRole('textbox')).toHaveValue('default');
  });

  it('should set defaultValue to value props when input was reset', () => {
    const Component = () => {
      const { reset, control } = useForm();

      React.useEffect(() => {
        reset({ test: 'default' });
      }, [reset]);

      return (
        <Controller
          defaultValue=""
          name="test"
          as={<input />}
          control={control}
        />
      );
    };

    render(<Component />);

    expect(screen.getByRole('textbox')).toHaveValue('default');
  });

  it('should render when registered field values are updated', () => {
    const Component = () => {
      const { control } = useForm();
      return (
        <Controller
          defaultValue=""
          name="test"
          as={<input />}
          control={control}
        />
      );
    };

    render(<Component />);

    fireEvent.input(screen.getByRole('textbox'), { target: { value: 'test' } });

    expect(screen.getByRole('textbox')).toHaveValue('test');
  });

  it("should trigger component's onChange method and invoke setValue method", () => {
    let fieldValues: any;
    const Component = () => {
      const { control, getValues } = useForm();

      return (
        <>
          <Controller
            defaultValue=""
            name="test"
            as={<input />}
            control={control}
          />
          {/**
           * We are checking if setValue method is invoked
           */}
          <button onClick={() => (fieldValues = getValues())}>getValues</button>
        </>
      );
    };

    render(<Component />);

    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: 'test' },
    });

    fireEvent.click(screen.getByRole('button', { name: /getValues/ }));

    expect(fieldValues).toEqual({ test: 'test' });
  });

  it("should trigger component's onChange method and invoke trigger method", async () => {
    let errors: any;
    const Component = () => {
      const { control, ...rest } = useForm({ mode: 'onChange' });

      errors = rest.errors;

      return (
        <Controller
          defaultValue="test"
          name="test"
          as={<input />}
          control={control}
          rules={{ required: true }}
        />
      );
    };

    render(<Component />);

    await act(async () => {
      fireEvent.input(screen.getByRole('textbox'), {
        target: { value: '' },
      });
    });

    expect(errors.test).toBeDefined();
  });

  it("should trigger component's onBlur method and invoke trigger method", async () => {
    let errors: any;
    const Component = () => {
      const { control, ...rest } = useForm({ mode: 'onBlur' });

      errors = rest.errors;

      return (
        <Controller
          defaultValue=""
          name="test"
          as={<input />}
          control={control}
          rules={{ required: true }}
        />
      );
    };

    render(<Component />);

    await act(async () => {
      fireEvent.blur(screen.getByRole('textbox'), {
        target: { value: '' },
      });
    });

    expect(errors.test).toBeDefined();
  });

  it('should set field to formState.touched', async () => {
    let touched: any;
    const Component = () => {
      const { control, formState } = useForm({ mode: 'onBlur' });

      touched = formState.touched;

      return (
        <Controller
          defaultValue=""
          name="test"
          as={<input />}
          control={control}
        />
      );
    };

    render(<Component />);

    await act(async () => {
      fireEvent.blur(screen.getByRole('textbox'));
    });

    expect(touched).toEqual({ test: true });
  });

  it('should call trigger method when revalidationMode is onBlur with blur event', async () => {
    const Component = () => {
      const { handleSubmit, control, errors } = useForm({
        reValidateMode: 'onBlur',
      });

      return (
        <form onSubmit={handleSubmit(() => {})}>
          <Controller
            defaultValue=""
            name="test"
            as={<input />}
            control={control}
            rules={{ required: true }}
          />
          {errors.test && <span role="alert">required</span>}
          <button>submit</button>
        </form>
      );
    };
    render(<Component />);

    fireEvent.blur(screen.getByRole('textbox'), {
      target: {
        value: '',
      },
    });

    expect(screen.queryByRole('alert')).not.toBeInTheDocument();

    await act(async () => {
      fireEvent.submit(screen.getByRole('button'));
    });

    act(() => {
      fireEvent.input(screen.getByRole('textbox'), {
        target: {
          value: 'test',
        },
      });
    });

    expect(screen.queryByRole('alert')).toBeInTheDocument();

    await act(async () => {
      fireEvent.blur(screen.getByRole('textbox'), {
        target: {
          value: 'test',
        },
      });
    });

    expect(screen.queryByRole('alert')).not.toBeInTheDocument();
  });

  it('should invoke custom event named method', () => {
    let fieldValues: any;
    const Component = () => {
      const { control, getValues } = useForm();
      return (
        <>
          <Controller
            defaultValue=""
            name="test"
            render={(props) => {
              return <input {...props} />;
            }}
            control={control}
          />
          {/**
           * We are checking if setValue method is invoked
           */}
          <button onClick={() => (fieldValues = getValues())}>getValues</button>
        </>
      );
    };

    render(<Component />);

    fireEvent.input(screen.getByRole('textbox'), {
      target: {
        value: 'test',
      },
    });

    fireEvent.click(screen.getByRole('button', { name: /getValues/ }));

    expect(fieldValues).toEqual({ test: 'test' });
  });

  it('should invoke custom onChange method', () => {
    const onChange = jest.fn();
    const Component = () => {
      const { control } = useForm();
      return (
        <>
          <Controller
            defaultValue=""
            name="test"
            render={({ onBlur, value }) => {
              return (
                <Input placeholder="test" {...{ onChange, onBlur, value }} />
              );
            }}
            control={control}
          />
        </>
      );
    };

    render(<Component />);

    fireEvent.input(screen.getByRole('textbox'), {
      target: {
        value: 'test',
      },
    });

    expect(onChange).toBeCalled();
  });

  it('should invoke custom onChange method', () => {
    const onBlur = jest.fn();
    const Component = () => {
      const { control } = useForm();
      return (
        <>
          <Controller
            defaultValue=""
            name="test"
            render={({ onChange, value }) => {
              return <Input {...{ onChange, onBlur, value }} />;
            }}
            control={control}
          />
        </>
      );
    };

    render(<Component />);

    fireEvent.blur(screen.getByRole('textbox'));

    expect(onBlur).toBeCalled();
  });

  it('should be null if as and render props are not given', () => {
    jest.spyOn(console, 'warn').mockImplementation(() => {});

    const Component = () => {
      const { control } = useForm();
      // @ts-ignore
      return <Controller defaultValue="" name="test" control={control} />;
    };

    const { container } = render(<Component />);

    expect(container).toEqual(document.createElement('div'));

    // @ts-ignore
    console.warn.mockRestore();
  });

  it('should update rules when rules gets updated', () => {
    let fieldsRef: any;
    const Component = ({ required = true }: { required?: boolean }) => {
      const { control } = useForm();
      fieldsRef = control.fieldsRef;
      return (
        <Controller
          defaultValue=""
          name="test"
          as={<input />}
          rules={{ required }}
          control={control}
        />
      );
    };
    const { rerender } = render(<Component />);

    rerender(<Component required={false} />);

    // @ts-ignore
    expect(fieldsRef.current.test.required).toBeFalsy();
  });

  it('should set initial state from unmount state', () => {
    const Component = ({ isHide }: { isHide?: boolean }) => {
      const { control } = useForm({ shouldUnregister: false });
      return isHide ? null : (
        <Controller
          defaultValue=""
          name="test"
          as={<input />}
          control={control}
        />
      );
    };

    const { rerender } = render(<Component />);

    fireEvent.input(screen.getByRole('textbox'), { target: { value: 'test' } });

    rerender(<Component isHide />);

    expect(screen.queryByRole('textbox')).not.toBeInTheDocument();

    rerender(<Component />);

    expect(screen.getByRole('textbox')).toHaveValue('test');
  });

  it('should not set initial state from unmount state when input is part of field array', () => {
    const Component = () => {
      const { control } = useForm({ shouldUnregister: false });
      const { fields, append, remove } = useFieldArray({
        name: 'test',
        control,
      });
      return (
        <form>
          {fields.map((field, i) => (
            <Controller
              key={field.id}
              defaultValue=""
              name={`test[${i}].value`}
              as={<input />}
              control={control}
            />
          ))}
          <button type="button" onClick={() => append({ value: 'test' })}>
            append
          </button>
          <button type="button" onClick={() => remove(0)}>
            remove
          </button>
        </form>
      );
    };

    render(<Component />);

    fireEvent.click(screen.getByRole('button', { name: /append/i }));

    fireEvent.input(screen.getByRole('textbox'), { target: { value: 'test' } });

    fireEvent.click(screen.getByRole('button', { name: /remove/i }));

    fireEvent.click(screen.getByRole('button', { name: /append/i }));

    expect(screen.getByRole('textbox')).not.toHaveValue();
  });

  describe('error handling', () => {
    it('should throw custom error if control is undefined in development environment', () => {
      jest.spyOn(console, 'error').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'development';

      const Component = () => {
        return <Controller as={'input' as const} name="test" defaultValue="" />;
      };

      expect(() => render(<Component />)).toThrow(
        '📋 Controller is missing `control` prop.',
      );

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.error.mockRestore();
    });

    it('should throw TypeError if control is undefined in production environment', () => {
      jest.spyOn(console, 'error').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'production';

      const Component = () => {
        return <Controller as={'input' as const} name="test" defaultValue="" />;
      };

      expect(() => render(<Component />)).toThrow(
        "Cannot read property 'control' of null",
      );

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.error.mockRestore();
    });

    it('should output error message if name is empty string in development environment', () => {
      jest.spyOn(console, 'warn').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'development';

      const Component = () => {
        const { control } = useForm();
        return (
          <Controller
            as={'input' as const}
            name=""
            control={control}
            defaultValue=""
          />
        );
      };

      render(<Component />);

      expect(console.warn).toBeCalledTimes(2);

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.warn.mockRestore();
    });

    it('should not output error message if name is empty string in production environment', () => {
      jest.spyOn(console, 'warn').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'production';

      const Component = () => {
        const { control } = useForm();
        return (
          <Controller
            as={'input' as const}
            name=""
            control={control}
            defaultValue=""
          />
        );
      };

      render(<Component />);

      expect(console.warn).not.toBeCalled();

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.warn.mockRestore();
    });

    it('should output error message if defaultValue is undefined in development environment', () => {
      jest.spyOn(console, 'warn').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'development';

      const Component = () => {
        const { control } = useForm();
        return (
          <Controller as={'input' as const} name="test" control={control} />
        );
      };

      render(<Component />);

      expect(console.warn).toBeCalledTimes(1);

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.warn.mockRestore();
    });

    it('should not output error message if defaultValue is undefined in production environment', () => {
      jest.spyOn(console, 'warn').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'production';

      const Component = () => {
        const { control } = useForm();
        return (
          <Controller as={'input' as const} name="test" control={control} />
        );
      };

      render(<Component />);

      expect(console.warn).not.toBeCalled();

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.warn.mockRestore();
    });

    it('should output error message if as and render props are given in development environment', () => {
      jest.spyOn(console, 'warn').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'development';

      const Component = () => {
        const { control } = useForm();
        return (
          // @ts-ignore
          <Controller
            as={'input' as const}
            render={() => <input />}
            defaultValue=""
            name="test"
            control={control}
          />
        );
      };

      render(<Component />);

      expect(console.warn).toBeCalledTimes(1);

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.warn.mockRestore();
    });

    it('should not output error message if as and render props are given in production environment', () => {
      jest.spyOn(console, 'warn').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'production';

      const Component = () => {
        const { control } = useForm();
        return (
          // @ts-ignore
          <Controller
            as={'input' as const}
            render={() => <input />}
            defaultValue=""
            name="test"
            control={control}
          />
        );
      };

      render(<Component />);

      expect(console.warn).not.toBeCalled();

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.warn.mockRestore();
    });

    it('should output error message if as and render props are not given in development environment', () => {
      jest.spyOn(console, 'warn').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'development';

      const Component = () => {
        const { control } = useForm();
        return (
          // @ts-ignore
          <Controller defaultValue="" name="test" control={control} />
        );
      };

      render(<Component />);

      expect(console.warn).toBeCalledTimes(1);

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.warn.mockRestore();
    });

    it('should not output error message if as and render props are not given in production environment', () => {
      jest.spyOn(console, 'warn').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'production';

      const Component = () => {
        const { control } = useForm();
        return (
          // @ts-ignore
          <Controller defaultValue="" name="test" control={control} />
        );
      };

      render(<Component />);

      expect(console.warn).not.toBeCalled();

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.warn.mockRestore();
    });

    it('should warn the user when defaultValue is missing with useFieldArray in development environment', () => {
      jest.spyOn(console, 'warn').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'development';

      const Component = () => {
        const { control } = useForm({
          defaultValues: {
            test: [{ data: '' }],
          },
        });
        const { fields } = useFieldArray({
          control,
          name: 'test',
        });

        return (
          <form>
            {fields.map(({ id }, index) => {
              return (
                <Controller
                  name={`test[${index}].data`}
                  as={'input' as const}
                  control={control}
                  key={id}
                />
              );
            })}
          </form>
        );
      };

      render(<Component />);

      expect(console.warn).toBeCalledTimes(1);

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.warn.mockRestore();
    });

    it('should not warn the user when defaultValue is missing with useFieldArray in production environment', () => {
      jest.spyOn(console, 'warn').mockImplementation(() => {});

      const env = process.env.NODE_ENV;
      process.env.NODE_ENV = 'production';

      const Component = () => {
        const { control } = useForm({
          defaultValues: {
            test: [{ data: '' }],
          },
        });
        const { fields } = useFieldArray({
          control,
          name: 'test',
        });

        return (
          <form>
            {fields.map(({ id }, index) => {
              return (
                // @ts-ignore
                <Controller
                  name={`test[${index}].data`}
                  control={control}
                  key={id}
                />
              );
            })}
          </form>
        );
      };

      render(<Component />);

      expect(console.warn).not.toBeCalled();

      process.env.NODE_ENV = env;

      // @ts-ignore
      console.warn.mockRestore();
    });
  });

  it('should not assign default value when field is removed with useFieldArray', () => {
    const Component = () => {
      const { control } = useForm();
      const { fields, append, remove } = useFieldArray({
        control,
        name: 'test',
      });

      return (
        <form>
          {fields.map((field, i) => (
            <div key={field.id}>
              <Controller
                as="input"
                name={`test[${i}].value`}
                defaultValue={''}
                control={control}
              />
              <button type="button" onClick={() => remove(i)}>
                remove{i}
              </button>
            </div>
          ))}
          <button type="button" onClick={() => append({ value: '' })}>
            append
          </button>
        </form>
      );
    };

    render(<Component />);

    fireEvent.click(screen.getByRole('button', { name: /append/i }));
    fireEvent.click(screen.getByRole('button', { name: /append/i }));
    fireEvent.click(screen.getByRole('button', { name: /append/i }));

    const inputs = screen.getAllByRole('textbox');

    fireEvent.input(inputs[0], {
      target: { value: '1' },
    });

    fireEvent.input(inputs[1], {
      target: { value: '2' },
    });

    fireEvent.input(inputs[2], {
      target: { value: '3' },
    });

    fireEvent.click(screen.getByRole('button', { name: /remove1/i }));

    expect(screen.getAllByRole('textbox')[0]).toHaveValue('1');
    expect(screen.getAllByRole('textbox')[1]).toHaveValue('3');
  });

  it('should validate input when input is touched and with onTouched mode', async () => {
    let currentErrors: any = {};
    const Component = () => {
      const { errors, control } = useForm<{ test: string }>({
        mode: 'onTouched',
      });

      currentErrors = errors;

      return (
        <form>
          <Controller
            name={'test'}
            control={control}
            defaultValue=""
            rules={{ required: true }}
            render={(props) => <input {...props} />}
          />
        </form>
      );
    };

    render(<Component />);

    const input = screen.getByRole('textbox');

    await act(async () => {
      fireEvent.blur(input);
    });

    expect(currentErrors.test).not.toBeUndefined();

    await act(async () => {
      fireEvent.input(input, {
        target: { value: '1' },
      });
    });

    expect(currentErrors.test).toBeUndefined();
  });
});
