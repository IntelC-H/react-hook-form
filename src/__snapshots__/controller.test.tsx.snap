// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controller should be assigned method.control variable if wrap with FormProvider 1`] = `
<DocumentFragment>
  <input
    name="test"
    value=""
  />
</DocumentFragment>
`;

exports[`Controller should render correctly with as with component 1`] = `
<DocumentFragment>
  <input
    name="test"
    value=""
  />
</DocumentFragment>
`;

exports[`Controller should render correctly with as with string 1`] = `
<DocumentFragment>
  <input
    name="test"
    value=""
  />
</DocumentFragment>
`;

exports[`Controller should support custom value name 1`] = `
<DocumentFragment>
  <input
    value="test"
  />
</DocumentFragment>
`;

exports[`Controller should support default value from hook form 1`] = `
<DocumentFragment>
  <input
    name="test"
    value="data"
  />
</DocumentFragment>
`;
